examples/chat2-reliable/chat.go

```diff
@@ -137,20 +151,65 @@ func (c *Chat) receiveMessages() {
 		case <-c.ctx.Done():
 			return
 		case value := <-c.C:
-
 			msgContentTopic := value.Message().ContentTopic
 			if msgContentTopic != c.options.ContentTopic {
 				continue // Discard messages from other topics
 			}
 
 			msg, err := decodeMessage(c.options.ContentTopic, value.Message())
 			if err == nil {
-				// send valid messages to the UI
-				c.ui.ChatMessage(int64(msg.Timestamp), msg.Nick, string(msg.Payload))
+				c.processReceivedMessage(msg)
 			}
 		}
 	}
 }
+
+func (c *Chat) processReceivedMessage(msg *pb.Message) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Update Lamport timestamp
+	if msg.LamportTimestamp > c.lamportTimestamp {
+		c.lamportTimestamp = msg.LamportTimestamp
+	}
+	c.lamportTimestamp++
+
```

Comment:
I think we can review the ACK status of messages in the `outgoingBuffer` here based on the received `bloomFilter` _and_ the received `causalHistory` (if our outgoing message is in the received causal history it can deterministically be marked as "sent" and probabilistically so if included in the `bloomFilter`).